### ISSUE TITLE
Atmos is way to slow

### ISSUE THREAD

**Issue by SpaceTimeNow** [ParadiseSS13/Paradise#5156]
**Problem Description**:
The Linda Atmos code is so slow that it practically is not able to calculate atmos in any reasonable amount of time.

**What did you expect to happen**: 
It being able to calc atmos in a reasonable amount of time.

**What happened instead**:
It's slow as fuck! I mean it's so slow that we have to drag portable srubbers or pump like vacuum cleaners accross every fucking tile we want to change the pressure or gas composition on. That's not an exaggeration, it's litterally what you have to do as an atmos tech!

**Why is this bad/What are the consequences:**
- Breaches or gas leaks have no real impact on the player unless you stand right on or next to a space tile
- It can take a good 30 minutes to vent or repressurize a room (depending on size and shape)
- Atmos Tech and Engineering jobs have become incredibly boring and redundant as result. 
- In general rounds have become way to tame, because enviromental danger is strongly reduced
- I could go on, but it should really be obvious that a functioning atmos system is kind of a big deal if you simulate a space station

**Suggestions to fix the problem**:
1. Scrap Linda and make something new
2. Find ways to improve Linda code
3. Make Linda process atmos more often
4. Code some fucked up solution where breaches suck x% (percentage reduced by distance) gas volume of every tile that is indirectly effected by it. (e.g. Tiles that are 20 tiles away but not seperated by a airlock or wall). Revesed on gas leaks. Sort of like the ZAS idea on top of Linda, so breaches and leaks will have a noticable effect trough the simplified ZAS idea and the smaller stuff like pushing pulling and equalizing will be done by the linda code.
5. Anything that works, PLEASE, just make it work somehow.

**Comment by Fox-McCloud** [comment]
Issue reports is for bugs or inconsistencies, not your personal platform to vent frustrations about things you don't like.

**Comment by SpaceTimeNow** [comment]
Only because you life in denail of it, doesn't make it go away

**Comment by LiquidNightfury** [comment]
Not a rare opinion but not a problem that can be solved with the available infrastructure. GG

**Comment by SpaceTimeNow** [comment]
Implementing such a crappy code and then just ignoring all the consequences is very dissapointing. If your atmos code cannot calc atmos in a reasonable amount of time its an fucking issue!

**Comment by SpaceTimeNow** [comment]
I also like how fast it got closed again, god forbid everyone comments on it

**Comment by LiquidNightfury** [comment]
Maybe you're new to this place. The LINDA talk is kind of a dead horse at this point. It's not going to have any significant change unless a coder comes along with proof that they can provide significant performance enhancements.

**Comment by SpaceTimeNow** [comment]
I told them that is a bad idea from day one and now i see shit like this: https://github.com/ParadiseSS13/Paradise/pull/5132 . It's insane!

**Comment by tigercat2000** [comment]
@SpaceTimeNow This is not a democracy. Your opinion of atmos means absolutely nothing unless you are willing to change it yourself, and all that guarantees is that you'll have a PR review which will most likely result in the pull being closed.

**Comment by SpaceTimeNow** [comment]
@tigercat2000 If your codebase dictatorship would have their priorties straight then this would have been fixed months ago. That i would waste my time coding for this codebase doesn't even register on scale of possibilities. There would have to be some major changes as to who runs it and how it's run.

**Comment by CrAzYPiLoT-SS13** [comment]
After reading your points, I seriously doubt in your reasoning. Atmos code is so fucked up no one is willing to touch it. Seemingly, including you. You cannot make an issue about something that is near impossible to do in the workings of BYOND and expect a proper reply. If you are willing to do something about atmos, feel free, just don't expect someone else to work their ass off to fix a non-broken, just hard to code thing. LINDA is a big improvement over ZAS. Take a look at the facts, don't just ramble on about it, doing nothing to change it.

**Comment by PJB3005** [comment]
> LINDA is a big improvement over ZAS.

Source pls.

**Comment by SpaceTimeNow** [comment]
@CrAzYPiLoT-SS13 The "facts" are that linda has to be artificially slowed down in order to not lag the server even more than ZAS. The whole conzept to replace ZAS (which processed areas as groups) with a system that calculates every fucking tile seperatly (Linda) is just insane. Linda ended up being such a resource hog (what a suprise) that the rate at wich chances happen is like 2 seconds and the changes that happen per tick are minimal.

To everyone that says that Linda is "working", that only applies to the system in theroy. The way it's implemented and configurated on our sever makes it incredibly slow performing. Performance is a big factor while coding and if your code cannot do the job it's supposed to do in a reasonable amount of time, which Linda obviously cannot, then it's still not practical solution. So who is just rambling about? 

The very fact that PRs like the atmos grenades exists, says everything!

And about the "code it yourself" part, which is a very weird arguement to make btw.: You must be delusional if you think that anyone would go through the effort and learn byond coding and implementing a new atmos system, if there is only a sliver of a chance that the code will just be denied and all your work goes to waste. That being said, I personally have neither the motivation nor the time to make that happen any time soon.

My issue is not so much about that it is no being fix. I would be totally ok if someone would finally acknowledge the issue and say: yeah, we know it sucks, but we don't have the people skilled or willing (or whatever) to do something about that. Just acknowledge the god damn issue and put a pin in it. We didn't even get that far, it's still being denied that this issue even exists! If we finally pass this point then there will be a chance that maybe someday someone will actually do something about that.

**Comment by PJB3005** [comment]
Every experience I've had with LINDA, it's been top of the profiler 24/7 and godawful performance where ZAS has a hard time even showing up on the profiler.

I'm not denying that Paradise's outdated version of ZAS performed well, but at the same point in time /vg/ long had ZAS optimizations, and for the most part we haven't underwent many optimizations since. They definitely should've just updated ZAS to /vg/ or bay versions and it'd be fine but no. They had to go the TG route because they can't maintain it themselves. TG has shown on *multiple* occasions to be shit at optimizing where needed and they'd rather beat dead horses instead.

The only issues I've ever had with ZAS were some zone forming issues and weirdness with vacuums exposed to space, both of which are relatively minor and I'd much rather have a good, fast atmos system than LINDA.

The fact that to get good siphoning speed LINDA needs to cheat by sucking a 3x3 area from a single machine just shows how obviously flawed the system is.

Yet TG is still goddamn delusional. TG made me realize that even in 2016 SS13 coding is still awful, while I had the impression it's been going way better when I was still sort of isolated to /vg/.

**Comment by Crazylemon64** [comment]
A hook in for the atmos algoriithm was proposed in the past, but the apparent majority of coders did not approve of it: #3214

**Comment by tigercat2000** [comment]
Don't dredge that old awful system up.

**Comment by SpaceTimeNow** [comment]
This and the fact that tg supposedly can run linda 4 times faster than us, kinda begs the question what kind of hardware this server is running on.

**Comment by PJB3005** [comment]
TG can run it that fast because of their master controller. It slows down the higher the CPU usage is.

**Comment by Spacemanspark** [comment]
Linda sped up like it is on /tg/ is more tedious than it is fun.
(Looking at you space wind)

**Comment by Fox-McCloud** [comment]
>(Looking at you space wind)

Hypothetically we could leave the current delay on space wind, but have rooms drain super fast and such.

That said, there would be a bit of work to be done first before we could do that (putting atmospheric machines on their own loop/in LINDA's loop); we can currently run pipenets and atmos at 4x their current speed, but the atmospherics machines (vents, pumps, etc) would still only run once every 2 seconds.

**Comment by SpaceTimeNow** [comment]
When vents are set to "refill" or what the setting is called, they will dump more than 113 kPa on a tile, do they really have to be running faster? Also couldn't they just dump double the amount of gas instead of ticking faster? From what i can observe ingame, the big issue seems to be that the pressure does not spread fast enough away from the vent tiles.

**Comment by Crazylemon64** [comment]
Simply increasing the clock rate of the atmos system won't cut it - with the nature of LINDA's algorithm, even mildly more distant tiles will not notice nearly as much impact than nearby ones, leading to a "jello-like" air system where tiles distant from a breach will not meaningfully notice(read: Lose/gain pressure/temperature from) a breach more than, say, 4 tiles away. This is most egregiously noticeable in maintenance corridors, where the 1-wide tunnels lead to a very slow air distribution, meaning that a hole in one end will take a LONG time for something even 5 tiles away to lose air - causing the air to move 4 times as often will make it happen 4 times as fast, but this process is exponentially slower for each tile further still, outpacing the linear rate increases we can do. While some prefer it this way, this means that hull breaches are ultimately not as significant than with an algorithm that allows gas to drain from tiles at a distance.

tl;dr: (3/4)^8 is a small number, and 4 * (3/4)^8 is still a small number.

**Comment by SpaceTimeNow** [comment]
@Crazylemon64
That is a good point. Was that somehow addressed in #3214?

The whole concept of this system is so weird to me.

**Comment by Crazylemon64** [comment]
#3214 caused tiles to be able to move air from a distance if there was enough of a difference between the two regions, fixing the above problem.

However: testing on my own side proves the complaints that the maintainers had.

The test: 
1. 10 devastation-5 explosions distributed across the station
2. Wait 5 seconds for deletes to finish up
3. Profile atmos movement for 5 minutes.

Normal LINDA:
Trial 1:
Air controller took 14.831 units of CPU time over the 5 minutes
Trial 2:
Air controller took 13.854 units of CPU time over the 5 minutes
Trial 3:
Air controller took 13.420 units of CPU time over the 5 minutes

Average: 13.076 units of CPU time

LINDAx4+Currents:
Trial 1:
Air controller took 30.558 units of CPU time over the 5 minutes
Air Currents controller took 8.042 units of CPU time over the 5 minutes
Sum: 38.600 units of CPU time
Trial 2:
Air controller took 21.832 units of CPU time over the 5 minutes
Air Currents controller took 7.281 units of CPU time over the 5 minutes
Sum: 29.113 units of CPU time
Trial 3:
Air controller took 29.841  units of CPU time over the 5 minutes
Air Currents controller took 8.735 units of CPU time over the 5 minutes
Sum: 38.576 units of CPU time

Average: 35.429 units of CPU time

The ratio of CPU time usage, between the two systems, is: x2.709.

NOTE: After further inspection, it appears that the code sample I was working with already upscaled LINDA to x4 processing rate. I'll turn it back down for parity with the regular rate, and examine the results then.
